neps/nep-0491.md

```diff
@@ -0,0 +1,359 @@
+---
+NEP: 491
+Title: Non-Refundable Storage Staking
+Authors: Jakob Meier <jakob@near.org>
+Status: Draft
+DiscussionsTo: https://gov.near.org/t/proposal-locking-account-storage-refunds-to-avoid-faucet-draining-attacks/34155
+Type: Protocol Track
+Version: 1.0.0
+Created: 2023-07-24
+LastUpdated: 2023-07-26
+---
+
+
+## Summary
+
+Non-refundable storage allows to create accounts with arbitrary state for users,
+without being susceptible to refund abuse.
+
+This is done by tracking non-refundable balance in a separate field of the
+account. This balance is only useful for storage staking and otherwise can be
+considered burned.
+
+
+## Motivation
+
+Creating new accounts on chain costs a gas fee and a storage staking fee. The
+more state is added to the account, the higher the storage staking fees. When
+deploying a contract on the account, it can quickly go above 1 NEAR per account.
+
+Some business models are okay with paying that fee for users upfront, just to
+get them onboarded. However, if a business does that today, their users can
+delete their new accounts and spend the tokens intended for storage staking in
+other ways. Since this is free for the user, they are financially incentivized
+to repeat this action for as long as the business has funds left in the faucet.
+
+The protocol should allow to create accounts in a way that is not susceptible to
+such refund abuse. This would at least change the incentives such that creating
+fake users is no longer profitable.
```

Comment:
An alternative solution to this problem that doesn't involve a change to the protocol.

The business creates an account with a deployed contract that will work in two modes. 
- Restricted mode: The user doesn't have a FullAccessKey, and can't delete the account
- Regular mode: The user has a FullAccessKey

On restricted mode, the user has a FunctionCallKey that allows calling into some methods of the contract that grants similar permissions to a FullAccessKey, but not all: [See all permissions here](https://docs.near.org/concepts/basics/accounts/access-keys#full-access-keys). In particular, it must not have permission to delete the account and get a refund.

To upgrade from restricted mode to regular mode, there is a method `upgrade` that will send the initial amount of NEAR deposited by the creator of the account (the business) and will attach a new FullAccessKey to the contract (the same as the initial FunctionCallKey). Anyone can call this method.

With this proposal, there will be no incentives to steal money from the business (by creating and deleting accounts), since the user can only delete the account in normal mode, after depositing the initial amount.

The business can upgrade accounts for those users that they already trust to not be acting maliciously. (i.e their goal is not to recover the money, but to avoid incentives for malicious users to deplete their funds).

## Drawback

The accounts in restricted mode don't behave in the same way as owning a regular account, since all actions need to be done through a special method (i.e. to do any Transfer or FunctionCall to another contract). 

This might be acceptable if the contract is a multi-sig (or similar interface) where the user is not expected to do Transfers or FunctionCalls directly anyway; or if the account is expected to be used on a specific application only, that is aware of this mode. 

To support using restricted accounts for arbitrary (allowed) use cases, wallets should be aware of it, which requires creating a standard about it. An alternative (that requires a protocol update), is creating a new type of key, in addition to FullAccessKey and FunctionCallKey, we can have GranularAccessKey, which works similarly to a FullAccessKey but doesn't have permission to do some actions (in this case it won't have permission to Delete the contract).

Reply:
cc. @jbajic

Reply:
> How does a user interact with a fungible token contract without a Full Access Key?

@DavidM-D  As I understand the idea, it goes through a contract deployed on the account that forwards arbitrary calls.

> On restricted mode, the user has a FunctionCallKey that allows calling into some methods of the contract that grants similar permissions to a FullAccessKey, but not all

So for example, you call `user.near::proxy([{action: "Function Call", "receiver": "ft.near", "method": "ft_transfer", "arg": "100", "balance": "1"}])` which will be unpacked by the contract and create the actual request as `ft.near::ft_transfer("100")` to the FT contract as a cross-contract call.

I think we discussed this as the "proxy account model" before, in the variation where it's deployed on every account. The increased storage and gas costs were one of the problems we saw in this idea. Also the latency is increased by at least one block time.

But more importantly, I think we assumed we want to give full access to users, which leads again to the storage refund problem. Now, the proxy contract gets around this problem by stripping the rights to delete the account or redeploy the contract code from the user. That's fine and I think we considered this already but assumed for the upgrade path we need to hold a full access key on the sponsor, which we didn't like for potential abuse issues. However, this seems to solve it:

> To upgrade from restricted mode to regular mode, there is a method upgrade that will send the initial amount of NEAR deposited by the creator of the account (the business) and will attach a new FullAccessKey to the contract (the same as the initial FunctionCallKey). Anyone can call this method.

So the upgrade path is also part of the contract. Nobody has full access for the account, except the contract code. I really like this idea by @mfornet 

Of course it's possible I missed something. WDYT think @DavidM-D, could it work for your use case?

Reply:
So to support the alternative solution, we would need to add another type of key (beforementioned **GranularAccessKey**) for this specific use case in which we don't want users to delete their account. I am a bit confused with this delegation of actions through FunctionCall, why would the restricted account have to use FunctionCall?

I am also fine with the existing solution, I feel like the alternative is slightly better but I may lack understanding to see if it is much better to pursue it instead of what we already have.

Reply:
@mfornet , @DavidM-D , @firatNEAR , shall we set up a meeting sometime early next week to sort it out so we can get to the alignment asap? EoY is right around the corner and we need to move fast

Reply:
I'm keen for that.

One question I'd ask is, ignoring the contract cost, could we have implemented Zero Balance Accounts entirely on the application level using a KeyPomesque solution?

Reply:
@walnut-the-cat I can join a call next week as well. Use the email: mfornet94 at gmail

> One question I'd ask is, ignoring the contract cost, could we have implemented Zero Balance Accounts entirely on the application level using a KeyPomesque solution?

@DavidM-D Zero Balance Accounts no, but maybe you could have achieved your goal anyway (which is not to avoid paying but avoid getting scammed by the users as I understand). 

How are the users expected to use those accounts, or interact with the contract you are deploying? (i.e what is the goal of the contract). 

If the contract is somehow "similar" in spirit to a multi-sig, where the user needs to go through a contract method to perform any action, then KeyPom contract will fit you. Otherwise **drawbacks** from the original comment in this thread applies.

Reply:
@mfornet , if you can share some of your availability (e.g. timezone), I will send out an invitation! I can check schedule for @DavidM-D and @firatNEAR from my end

Reply:
@walnut-the-cat I can do any time between (6:00 - 14:00) UTC, I could do later, but not every day. I recommend you create an event, and we follow up in email.

Reply:
What is the conclusion of this discussion? @mfornet @DavidM-D @jakmeier